src/irmin-containers/linked_log.mli

```diff
@@ -0,0 +1,83 @@
+(*
+ * Copyright (c) 2020 KC Sivaramakrishnan <kc@kcsrk.info>
+ * Copyright (c) 2020 Anirudh Sunder Raj <anirudh6626@gmail.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+(** [Linked_log] is a linked list implementation of the log. In this, two copies
+    of the log share their common predecessor. The type of the log entries, the
+    method to obtain timestamp, a hashing method and a content addressable store
+    maker are provided by the user. *)
+
+(** Signature of [Linked_log] *)
+module type S = sig
+  module Store : Irmin.S
+  (** Store for the log. All store related operations like branching, cloning,
+      merging, etc are done through this module. *)
+
+  type value
+  (** Type of log entry *)
+
+  type cursor
+  (** Type of cursor. Cursor is like a marker from which a certain number of
+      entries can be read *)
+
+  val append :
+    info:Irmin.Info.f -> Store.t -> path:Store.key -> value -> unit Lwt.t
+  (** Append an entry to the log *)
+
+  val get_cursor : Store.t -> path:Store.key -> cursor option Lwt.t
+  (** Get the cursor *)
+
+  val read : cursor -> num_items:int -> (value list * cursor option) Lwt.t
+  (** Read a certain number of entries starting from the cursor. If the number
+      specified is greater than the number of log entires from the cursor, the
+      log is read till the end. If the input cursor has already reached the end,
+      then an empty list is returned *)
+
+  val read_all : Store.t -> path:Store.key -> value list Lwt.t
+  (** Read the entire log *)
+end
+
+(** [Make] returns a mergeable linked log using the backend and other parameters
+    as specified by the user. *)
+module Make
+    (Backend : Irmin.S_MAKER)
+    (M : Irmin.Metadata.S)
+    (P : Irmin.Path.S)
+    (B : Irmin.Branch.S)
+    (H : Irmin.Hash.S)
+    (C : Cas_maker.S)
+    (T : Time.S)
+    (K : Irmin.Hash.S)
+    (V : Irmin.Type.S) :
+  S with type value = V.t and type Store.key = P.t and type Store.branch = B.t
+
+(** Linked log instantiated using {{!Irmin_unix.FS} FS backend} provided by
+    [Irmin_unix], timestamp method {!Time.Unix}, hash {!Irmin.Hash.SHA1} and CAS
+    maker {!Cas_maker.Mem} *)
```

Comment:
I wanted to make it easier for the user by providing a content addressable store. In order to do that, I also had to provide the corresponding config as I did not find a suitable way to take it from the user. `Irmin_git.config` requires me to provide a root. `Irmin_mem.config` did not require any such root. As I did not want to specify any root, I opted for the in-memory store. Passing a config was the hindrance which made me opt for in-memory store.
Possible fix: Is it acceptable to pass a root like `/tmp` to the config and provide that config for the content addressable store?

Reply:
I think it's not possible to provide a sensible default config for most stores (not all disk partitions / networks are equal). Even ignoring that, I think it's good to force the user to reason about store location / system resources whenever they want something other than an in-memory store. (If they don't care about where the data is going or retrieving it later, why not just use an in-memory store?)

I'm also not keen on having _every_ log created by a given module be linked implicitly via internal mutable state. How about we kill two birds with one stone and re-think the API to something with an explicit notion of sharing?

```ocaml
module type S = sig
  type t
  (** The type of linked logs. — internally, something like [Store.t * pool] *)

  type pool
  (** The type of pools of linked log instances. — internally, content-addressable store *)

  val create_pool : Irmin.config -> pool

  val v : Store.t -> pool -> t
  (** Create a new linked log that shares its history with the given pool of instances. *)

  (* Perhaps offer a [t -> Store.t] to allow underlying stores to be recovered 
   * from linked logs... Not strictly necessary *)
end
```

If you really want to keep the implicit sharing, I suggest adding functors to `Cas_maker` that take user-provided configs for stores like Unix without a sensible default. (At that point, it would also be worth making `Linked_log.Make` generative to make sharing more explicit.)